Update readme with release and deployment instructions. (#1737)
USACE/cwms-data-api: RELEASE_DEPLOY.md

Reviewer: It seems the regex's  for tags (dev,test,prod) might allow overlap.   For example a test tag could also pass as a release tag.    Consider if the prod tag should _not_ have the word 'test' or 'dev',   or perhaps **prod** tag should have 'prod' in the name?

Author: dev and test must have `-dev` or `-test` at the end, prod requires no such element, just allows for the [a-z]

# Releases and Deployments

(NOTE: except for dev/develop) this procedure is not in place, at least for deployments. Up to test will work, but is not active. The

> current process is not easily shared and is getting replace within the next few months so we will not spend time documenting them.

Reviewer: ```suggestion
current process is not easily shared and is getting replaced within the next few months so we will not spend time documenting them.
```

> However, developers authorized to generated a release that should go to one of the environents can still use this process to prepare

Reviewer: ```suggestion
However, developers authorized to generate a release that should go to one of the environments can still use this process to prepare
```

release for that manual deployment.

Releases are anything publish to the (releases)[https://github.com/usace/cwms-data-api/releases] page and can be uses for a local 
deployment or testing.

Deployments are specific to USACE's official CWMS dev,test, and prod environments and are controlled by authorized USACE staff.
Additionally deployments are made from a specific release, except for the dev environment and develop branch a deployment cannot be 
made without a release.

> For deployments there are github action "environemnts" that hold appropriate secrets and properties for values specific to that given

Reviewer: ```suggestion
For deployments there are github action "environments" that hold appropriate secrets and properties for values specific to that given
```

environment. Github actions that perform deployments are environment aware and will use them as required.

## Naming

> Calendar Version is used in this project. Previous attempts are SemVer ened up, to say the least, non-sensical. There were

Reviewer: ```suggestion
Calendar Version is used in this project. Previous attempts with SemVer ended up, to say the least, non-sensical. There were
```

too many changes too fast, and continue to be. Additionally the database schema backing the API uses Calender Versioning as well 
and the use of the same here makes it more obivious how out-of-sync they may be.

Releases and deployments are based on tags. Tags must follow the following formatting (regular expressions are used for variable 
portions):

| environment | tag/branch  | description |
| ------------ | ---- | ----- |

> | dev          | tag -> `\d\d\d\d\.\d\d\.\d\d-dev[a-z]+` | Tagged can be deployed to the dev environment if desired. That deployment will be overridden at night. These tags are most useful for downstream testing |

Reviewer: overridden by the nightly build?

Author: yeah, that would be better language

1. A war file available under the (releases)[https://github.com/usace/cwms-data-api/releases] page
2. OCI container images available at `docker pull ghcr.io/usace/cwms-data-api:<tag>`
    1. Note that there may be multiple tags for the same image, use what is appropriate to your current operation.

### Creating a Release

The project uses trunk based development. All releases are directly from the develop branch, *except* for specific hotfixes on an
as-needed basis. More information available in the deployment section

Tag as follows:

1. A given version is behaving in dev and it should be in test
2. Tag that commit with the appropriate test tag.
3. Once that version has shown to behave in the test environment, tag that *same* commit with the appropriate release tag

To create a release perform the following steps

1. Go to the (releases)[https://github.com/usace/cwms-data-api/releases] and draft a new release (unless directly releasing or deploying) the `develop` target, that release already exists
1. If deploying/releasing develop directly, skip the following steps. The process is different.
2. Select "draft new release"

> 3. If tagging the current state of develop follow github release prodecures to create a tag in the CalVer format and select "create new tag" and use the same name for the release as the tag.

Reviewer: ```suggestion
3. If tagging the current state of develop follow github release procedures to create a tag in the CalVer format and select "create new tag" and use the same name for the release as the tag.
```

4. Publish release. Unless a prod release, check the pre-release box.
5. NOTE: the details will be filled in by the triggered github actions.

#### Releasing "develop"

Go to the repositories "actions" tab and find the Nightly Releases - Build action. Click "Run workflow" and accept the defaults:

![Image of ](.readme-images/ActionsTab-Release%20Develop.png)

# Deployments

## Deploying Develop

Deployment of develop can be done without the initial release step. The process is the same as generating a release for develop except

> that the Nightly Releases - Schedule action is used. This action will build the release and they deploy it.

Reviewer: ```suggestion
that the Nightly Releases - Schedule action is used. This action will build the release and then deploy it.
```

## Deployments in general

Deployments in the various environment are expected to work. However, Each environment may be targetting a different database schema
version. 

> The projects automated tests target the latest developemnt version of the database, the current release version of the database, and

Reviewer: ```suggestion
The projects automated tests target the latest development version of the database, the current release version of the database, and
```
